### ISSUE TITLE
Latest App Store Release broke "Open in..." functionality?

### ISSUE THREAD

**Issue by mchalkley** [terrychou/iVim#185]
Hi Terry,

Since the iVim App Store update a couple days ago, I can't get to the "Open in..." functionality anymore.  I use it very frequently, especially from an app called Notebooks 10.  iPadOS 13 changed it somehow so that I had to tap-hold on iVim from the Share Sheet to get it to show up, but now I can't seem to get to it via any means at all.  Could this be because of the changes you made to the external documents code?

Thanks,

Mark

**Comment by terrychou** [comment]
Hi, Mark

1) By "can't seem to get to it via any means at all", do you mean that "iVim" disappear from the Share Sheet, or the "tap-hold" trick stops working?

2) Is this specific to `iVim`, or does it happen to any other apps?

3) Have you tried the share functionality from within `Files` app, or via the command `:idocuments`?

4) Have you tried to restart the device?


Thanks for the feedback. Yes, the latest update has made big changes to the external documents opening. `iVim` does not cache or make mirroring files to external documents anymore. Instead, it now opens documents at their original locations, which makes the process more efficient. On the other hand, this would undoubtedly bring some problems or inconvenience. But let's figure it out and fix it.

**Comment by mchalkley** [comment]
Hi Terry,

Thanks for the questions for clarification.

1.  The option for "Open in..." has disappeared completely.
2.  It still shows up with the other apps I've tried it with (that worked before, at least).
3. I am able to open the files belonging to Notebooks by using :idoc, so it does work - it's just too tedious to use regularly, due to the folder navigation issues...
4. I have tried restarting the iPad.
Thanks again - I know you'll figure it out - you always do - and I'll be glad to help in any way I can,
Mark

**Comment by terrychou** [comment]
`iVim` completely disappears from the Share Sheet? Hmm.. weird...

Can you please try the Share Sheet in the `Files` app? See whether the problem is there too?

And can you please try another app other than Notebooks? Because I don't have it installed, and I need to reproduce the problem on my device to test.

**Comment by mchalkley** [comment]
I'm sorry for the confusion, Terry...  iVim is available in the Share Sheet as "Share with app...", but not as "Open in..."  This is true for the Files app, as well.  

Until the last release of iVim, if you opened the Share Sheet, then tap-hold on iVim, you also had an "Open in..." option, which is what I use.  It allows you to edit an apps files in place, with the changes automatically being applied to the app's file.  

In other words, I can have text or Markdown documents in my Notebooks reference system, but make large changes to them in iVim very easily.

Mark

**Comment by terrychou** [comment]
The reason I keep asking you for trying other apps other than `Notebooks` is that on my end, the `Open in ...` item of `iVim` is still available and works. If it is the same case for other apps as for `Notebooks` on your end, I suspect this may be a problem similar to issue #76 .

I will have a deeper look into it, and am considering removing the text sharing part.

**Comment by mchalkley** [comment]
As an experiment, in opened the Files app, navigated to a text file, tap-hold on it, tap Share, tap-hold on iVim, only option is "Share with app".  Repeat the same process, but tap-hold on Drafts, 2 options show up: "Share with app" & "Open in app".  Repeat the same process with GoCoEdit, 1 option shows up: "Open in app".  If there are any other text editors or similar apps that you have that you'd like me to test with, let me know...

Thanks again,

Mark

**Comment by terrychou** [comment]
Sorry if my words were confusing. By "other apps", I mean if you can try and see whether the `iVim` sharing item works normally for other apps instead of `Notebooks`. In that case, I can tell if it is a problem specific to `Notebooks`, or to the whole system.

**Comment by mchalkley** [comment]
Yes, it appears to be the whole system - none of the apps I've tried can access the "Open in" functionality for iVim, whereas it's still working with other apps that had that functionality.

**Comment by terrychou** [comment]
Hi, Mark

I have made some changes to `iVim` that may be concerning this problem. Would you like to help and test it? If so, I will push it onto TestFlight for you.

**Comment by mchalkley** [comment]
Yes, absolutely, Terry - thanks!

**Comment by terrychou** [comment]
> Yes, absolutely, Terry - thanks!

Thanks, Mark. I have uploaded the app to TestFlight. However, it needs reviewing to be published for testing, which is unavailable during the holiday. I will push it to you as soon as it becomes available.

**Comment by mchalkley** [comment]
Just got it, Terry.  Preliminary indications are: IT WORKS!!!  Thank you!!!

Mark

**Comment by terrychou** [comment]
@mchalkley Great! Please give it more testings. If it works well, I will include this in the next update.

**Comment by mchalkley** [comment]
@terrychou So far, I've tested it on numerous files from Notebooks & Files, and it's working perfectly.  I'll continue testing.

Mark

**Comment by mchalkley** [comment]
@terrychou I've noticed some swap file errors that I wasn't getting before - not sure if it's related to this change or not...  I believe it resulted in the loss of a single line of text on one occasion, but I can't be sure, since it was only once.  Still testing...

**Comment by mchalkley** [comment]
@terrychou I just had a situation where I used "Open in" iVim and got a blank file with the correct name.  I then tried ":so%" to see what would happen and got an E484 "couldn't open <file name>".  Deleted the buffer (didn't have to override), re-tried the "Open in" and it was fine.

**Comment by terrychou** [comment]
> @terrychou I've noticed some swap file errors that I wasn't getting before - not sure if it's related to this change or not...  I believe it resulted in the loss of a single line of text on one occasion, but I can't be sure, since it was only once.  Still testing...

@mchalkley What kind of file did you open in those situations? Did you open them as single files, or as a directory first then open the files in netrw?

**Comment by terrychou** [comment]
> @terrychou I just had a situation where I used "Open in" iVim and got a blank file with the correct name.  I then tried ":so%" to see what would happen and got an E484 "couldn't open <file name>".  Deleted the buffer (didn't have to override), re-tried the "Open in" and it was fine.

@mchalkley what was the file type (e.g. file extension name) in this situation?

**Comment by mchalkley** [comment]
@terrychou In all cases, I was opening a single .md (markdown text) file.

**Comment by terrychou** [comment]
@mchalkley I have pushed version 2.28(2) to TestFlight. It includes to notify in some cases of external item opening failures, which may be helpful to provide useful information.

**Comment by mchalkley** [comment]
@terrychou Got it - thanks - I'll let you know how the testing goes.

**Comment by terrychou** [comment]
@mchalkley I have just pushed version 2.28(4) to App Store. Let me know if you have any further problem.

**Comment by mchalkley** [comment]
@terrychou I haven't noticed any occurrences of this issue.  Occasionally, when using the "Open in" functionality, I have to do it twice before the text appears in the iVim buffer.  This is only a very minor annoyance, and I can't reproduce it with any degree of consistency, so I think this issue can be closed.

**Comment by terrychou** [comment]
@mchalkley I have encountered the glitch you are mentioning too and I have some ideas concerning it. I'll try some fixings and push it to TestFlight for you to test when ready.

**Comment by mchalkley** [comment]
Any time - I love testing iVim!  (Even though it does usually just involve just doing what I usually do....) 😄

**Comment by terrychou** [comment]
@mchalkley I've pushed version 2.29(1) to TestFlight, which should fix this glitch and probably also issue #184 . It is under previewing right now and should be available soon.

**Comment by mchalkley** [comment]
I haven't see any empty buffer issues using "Open in" with the latest TestFlight beta, but I did experience a crash this morning when using "Open in".  That's the first iVim crash I've experienced in quite a while.  I haven't been able to reproduce it, and wasn't doing anything unusual that I'm aware of...

**Comment by terrychou** [comment]
I think I know the probable cause of the crash. I will make some changes and upload the update to TestFlight when it is ready.
